test/spec/ol/layer/layer.test.js

```diff
@@ -602,6 +602,29 @@ describe('ol.layer.Layer', function() {
       map.dispatchEvent(new ol.render.Event('precompose', map, null,
           frameState, null, null));
       expect(frameState.layerStatesArray.length).to.be(0);
+
+      layer.setMap(null);
+      map.dispatchEvent(new ol.render.Event('precompose', map, null,
+          frameState, null, null));
+      expect(frameState.layerStatesArray.length).to.be(0);
+
+      layer.setMap(map);
+      map.dispatchEvent(new ol.render.Event('precompose', map, null,
+          frameState, null, null));
+
+      expect(frameState.layerStatesArray.length).to.be(1);
+      layerState = frameState.layerStatesArray[0];
+      expect(layerState.layer).to.equal(layer);
+      expect(frameState.layerStates[goog.getUid(layer)]).to.equal(layerState);
+
+      layer.setMap(map);
+      map.dispatchEvent(new ol.render.Event('precompose', map, null,
+          frameState, null, null));
+
+      expect(frameState.layerStatesArray.length).to.be(2);
+      layerState = frameState.layerStatesArray[1];
+      expect(layerState.layer).to.equal(layer);
+      expect(frameState.layerStates[goog.getUid(layer)]).to.equal(layerState);
```

Comment:
I think we should separate out things and add new tests (i.e. new `it` functions) for the new use-cases under tests.